### PR TITLE
add initial one-pager for stack ui metadata

### LINKED ISSUE #544: Package/Resource UI configuration metadata
**Problem:**
Currently, if you are generating a CRD via a UI you have to hard code the specific UI configuration fields for a resource.

**Proposal:**
Adding a UI configuration file to the initial package file that will be associated to the initial spec and also allow for extending the validation requirements for given fields.

This will allow someone to drive the unique configuration fields displayed in the UI from their initial package file and thus no longer require any sort of hard coded UI elements on the frontend.

### PR DESCRIPTION
**Pull request by displague**
Add an initial one-pager draft for how Stacks should bundle UI
annotations.

This document was initially drafted by @rathpc out of tree.

Closes #544 

[skip ci]

### CODE REVIEW

File: design/one-pager-stack-ui-metadata.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+* Owner: Steven Rathbauer (@rathpc), Marques Johansson (@displague)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Terms
+
+* **Stack**: A Stack is a Crossplane extension-manager managed package.  These packages may bundle one or many Crossplane applications or infrastructure provider controllers.
+* **Stacks**: Stacks could refer to more than one Crossplane extension-manager packages or the Extension-Manager system as a whole.
+
+See the [Stacks Design Doc](design-doc-stacks.md) for more details about Crossplane Stacks.
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: this PR should probably also update the main design doc with a short description of this new UI focused addition to the stack package format

File: design/one-pager-stack-ui-metadata.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+* Advanced validation (beyond what Kubernetes' OAS3 support delivers)
+* ncurses-style installers
+* web-based package managers (out-of-cluster and in-cluster)
+* web-based configuration panels
+
+This spec is not concerned with a specific UI implementation but serves to provide a foundation for supporting Stack developers and users through basic form input.
+
+## Proposed File format
+
+The `ui-schema.yaml` file will conform to a specification that will not be covered formally in this document but will be suggested through examples and supporting text.
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: this sentence could use a bit more explanation around why we are not covering a formal specification, nor linking to an external formal specification.  The reason we aren't doing that is because we do not want to be prescriptive or authoritative about the capabilities of any UI that could be driven from this content, we want to be neutral and agnostic about this.

File: design/one-pager-stack-ui-metadata.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+## Proposed UI Elements
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: This section could use a sentence explaining that these are possible or potential UI elements, but not mandated  or required, and other UI elements not listed here are possible too.

File: design/one-pager-stack-ui-metadata.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+The primary purpose of the `ui-schema.yaml` format is to allow for easy authoring of the definition of UI markup, validation, and errors for a more complete UI and UX. The `ui-schema.yaml` file will be parsed and validated as part of package validation and ultimately serialized as a `JSON` annotation on the respective CRD at install time.
+
+### Example addition of `ui-schema.yaml` in a package tree
+
+```
+.registry/
+├── icon.png
+├── app.yaml # Application metadata.
+├── ui-schema.yaml # Optional UI spec for configuration metadata
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: what does a `ui-schema.yaml` file at the root mean vs. one that is scoped to a single CRD?  How would one at the root be processed and consumed?

File: design/one-pager-stack-ui-metadata.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+### CRD Annotation Example
+
+An example injection of the `ui-schema.yaml` as a CRD annotation would present as follows:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: is there a max length limit for annotations on objects in the k8s api?

File: design/one-pager-stack-ui-metadata.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+.
+.
+.
+.
+.
+```
+
+## Formal Specification
+
+TBD: A specification for the supported elements and their properties.  A validator and validation document pair should be offered.  There is no clear YAML validation format as there is DTD for XML documents.  Validation may be deferred to the capabilities of YAML parsing in Go via <https://github.com/go-yaml/yaml/>.  Unrecognized fields will be ignored while typed parameters will expect to conform to a Go type.
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: This is probably out of scope for Crossplane's stack manager, I think we'd like to be agnostic around any particular UI schema spec.  We likely will transparently take the given schema in a Stack and apply it as an annotation to the CRDs from the Stack that we create as opposed to dictating anything about the schema that we'll allow.

However, the general information about validation that we can do is still useful.

File: design/one-pager-stack-ui-metadata.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+## Open Questions
+
+* Should alternate UI schema be supported?
+  * Should this be provided as multiple `ui-schema.yaml` files (with different names) or embedded in a single file?
+    * Should a single CRD annotations include the nested or alternative schemas?
+    * Should multiple CRD annotations represent alternate schemas? How does this affect the annotation key?
+* Should dependent and co-dependent fields be supported?
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: I think this is out of scope for Crossplane as well if we are unopinionated about the UI schema itself.

File: design/one-pager-stack-ui-metadata.md (hunk 5 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: 3 years ago it increased from 64k to 256k

https://github.com/kubernetes/kubernetes/blob/0b4275b6c7ecd55cfd9916dbef8fe48ae4a8c733/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L35

File: design/one-pager-stack-ui-metadata.md (hunk 2 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: That's fair.  My reasoning for adding this was to the avoid creating a schema file because I couldn't find an appropriate validation spec or tool.  I assumed we could reasonably validate on some field names and values without blocking additions.

If the intent is to leave the format wide open without any meaningful validation, that does simplify things, including the implementation:  `yaml` -> `json` -> annotatation.

File: design/one-pager-stack-ui-metadata.md (hunk 4 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: I've wondered if there is may be reasons to annotate the Extension itself in addition to the extension created CRD, for visual package-configuration aids (like debconf).

@rathpc, what did you have in mind here?

File: design/one-pager-stack-ui-metadata.md (hunk 6 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Are you suggesting that valid yaml is the only requirement before passing the file along for annotation?  This is fine by me.

In chats the idea has been expressed that this format should allow for multiple ui-schemas to be defined, competing or nesting.  If we don't define any fields, ancillary filenames, or naming formats as ascribing special meaning, we leave this concerns to be figured out by the community at a later time.  This is also fine.

File: design/one-pager-stack-ui-metadata.md (hunk 4 above)
Comment (rathpc) [reply design/one-pager-stack-ui-metadata.md]: This is a valid question. The inclusion of these different specs was originally added by @lukeweber and I believe the intention was to have a fallback during the consumption/overlay process that creates the annotation. If a specific CRD version does not contain the `ui-schema.yaml` file it would look for a root fallback to consume instead. This is really only necessary if a CRD has a mandatory requirement for a ui field, version agnostic. That is why it is also listed as optional.

Let me know if you need further clarification on this or if you guys have other thoughts. Thanks!

File: design/one-pager-stack-ui-metadata.md (hunk 1 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Added in 13b4da9

File: design/one-pager-stack-ui-metadata.md (hunk 4 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: I added this reasoning in 13b4da9 ..  I'm still wondering if Extensions themselves (and not the CRDs they provide/manage) may be able to take advantage of UI markup. (Configuring your Provider's credentials in an Infra Stack, for example).

File: design/one-pager-stack-ui-metadata.md (hunk 5 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Noted this in 13b4da9

File: design/one-pager-stack-ui-metadata.md (hunk 6 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Expanded on this section in 13b4da9

File: design/one-pager-stack-ui-metadata.md (hunk 2 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Explained in 418e3ab4041f038ba50eac0477bea30fe1cac494

File: design/one-pager-stack-ui-metadata.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+* Advanced validation (beyond what Kubernetes' OAS3 support delivers)
+* ncurses-style installers
+* web-based package managers (out-of-cluster and in-cluster)
+* web-based configuration panels
+
+This spec is not concerned with a specific UI implementation but serves to provide a foundation for supporting Stack developers and users through basic form input.
+
+## Proposed File format
+
+The `ui-schema.yaml` content will not be dictated in this design doc.  Rather the [formal specification](formal-specification) is a simple guideline for the type and size of the document.  The contents are otherwise left open to interprettation by Stack tools with the expectation that these tools will promote interoperable standards.  Specifications suggested through examples and supporting text in this design document should not be considered doctrine.
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: this link is broken, probably just missing the `#` anchor hash?

Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: typo: `interprettation` -> `interpretation`

File: design/one-pager-stack-ui-metadata.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+.
+.
+.
+```
+
+## Formal Specification
+
+The current design calls for unbiased processing of the UI YAML to annotations, regardless of content.
+
+Because the YAML will be transcribed as JSON in the annotation:
```
Comment (jbw976) [inline design/one-pager-stack-ui-metadata.md]: something i hadn't really thought about until now: the stack package author writes the UI markup content in yaml, but transcribe (re-encode) into JSON to store as an annotation on the CRD?

Does it make more sense to expect the stack author to initially write their UI markup in JSON, OR to store the UI markup in the annotation as YAML?

Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: JSON is valid YAML: https://en.wikipedia.org/wiki/YAML#Comparison_with_JSON (some restrictions apply)

For hand-editing, I think escaping YAML within YAML may be more intricate that doing so with JSON. 
For tool digestion, I don't think it matters if the expression of UI in the Kubernetes object is JSON or YAML.
For reviewing Kubernetes objects and their properties by sight, I might prefer YAML.

If we use a YAML parser instead of a JSON parser when interpreting these fields we can accommodate both.  I have what is probably a misconception that nested or embedded YAML property values have some problems.  Other than the sense of some forgotten readings, my only evidence is that JSON is often embedded.

Let's aim for YAML.

---

@suskin asked yesterday why we even need an `app.yaml`  and `ExtensionRequest` when the `Extension` (and its `status`) could serve as both.  Working on inertia, I couldn't offer a reason.

**Comment by displague** [comment]
Let me know if you want the others squashed (or feel free to use the Github Squash+Merge). 

There's no past like the present.

File: design/one-pager-stack-ui-metadata.md (hunk 9 above)
Comment (displague) [reply design/one-pager-stack-ui-metadata.md]: Let's revisit this in #604
